aten/src/ATen/native/quantized/cudnn/Linear.cpp

```diff
@@ -0,0 +1,345 @@
+// Multiplication ops: rhs_mult_op, requant_op
+// Addition op: add_op
+// Relu op: relu_op
+template <bool kReluFused>
+void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_output, const at::Tensor& input, double output_scale) {
+  if (quantized_output.numel() == 0) {
+    return;
+  }
+  at::Tensor linear_output = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  // TODO: combine empty & fill_ using full_like or full
+  at::Tensor requantize_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  auto act_scale = input.q_scale();
+  auto weight_scale = orig_weight.q_scale();
+  auto requantize_multiplier = act_scale * weight_scale / output_scale;
+  requantize_multiplier_tensor.fill_(requantize_multiplier);
+  c10::optional<at::Tensor> bias_multiplier_tensor;
+  c10::optional<at::Tensor> broadcasted_bias;
+  if (bias_.has_value()) {
+    // the input bias is a 1-D tensor whose size is the same as the size of the second dimension of quantized_output.
+    // we need to add trailing dimensions in order to properly broadcast bias, otherwise broadcast_to will fail.
+    // the number of trailling dimensions is quantized_output.dim() - 2, so the new size of the broadcast_bias
+    // becomes quantized_output.dim() - 2 + 1. nothing needs to be done for the leading dimensions
+    std::vector<int64_t> new_size(quantized_output.dim() - 1, 1);
+    new_size[0] = bias_.value().size(0);
+    broadcasted_bias = bias_.value().reshape(new_size);
+    broadcasted_bias.value() = broadcasted_bias.value().broadcast_to(quantized_output.sizes());
+    bias_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+    auto bias_multiplier = 1.0 / (act_scale * weight_scale);
+    bias_multiplier_tensor.value().fill_(bias_multiplier);
+  }
+
+  cudnnHandle_t handle = at::native::getCudnnHandle();
+  CacheKey key;
+  bool deterministic{true};
+  bool allow_tf32{false};
+
+  key.input_alignment = cudnn_utils::getAlignment(input);
+  key.output_alignment = cudnn_utils::getAlignment(linear_output);
+  key.weight_alignment = cudnn_utils::getAlignment(orig_weight);
+  if (bias_.has_value()) {
+    key.bias_alignment = cudnn_utils::getAlignment(broadcasted_bias.value());
+  } else {
+    key.bias_alignment = -1;
+  }
+  // cudnn expects tensors to be at least 3D. act is currently 2D. we will create a 3D view
+  // cudnn expects leading dimensions to be the dummy dimensions
+  std::vector<int64_t> new_sizes(3, 1);
+  new_sizes.back() = orig_weight.size(0);
+  new_sizes[1] = orig_weight.size(1);
+  auto weight_transposed = transpose(orig_weight, 0, 1).view(new_sizes);
+  // TODO: remove this with int8 matmul is supported
+  auto input_fp = input.int_repr().to(at::kFloat);
+  auto weight_fp = weight_transposed.int_repr().to(at::kFloat);
+
+  auto run = [&](cudnn_frontend::ManagedOpaqueDescriptor plan_desc) {
+    auto workspace_size = 0;
```

Comment:
I feel in general we have a lot of bolierplate code, maybe we can think about creating some helper functions or easier abstractions to make this simpler, this will be helpful when we have more ops in cudnn